### ISSUE TITLE
bug(2048): action_mask calculated before adding random cell

### ISSUE THREAD

**Issue by aar65537** [instadeepai/jumanji#127]
### Description

The action mask returned by Game2048 is calculated before adding the random cell, which leads to incorrect action masks.
```
env = Game2048()
state = State(
    board=jnp.array(
        [[0, 1, 2, 3], [3, 1, 2, 3], [1, 2, 3, 4], [4, 3, 2, 1]], dtype=jnp.int32
    ),
    step_count=jnp.array(0, dtype=jnp.int32),
    action_mask=jnp.array([True, False, True, True], dtype=jnp.bool_),
    score=jnp.array(0.0, dtype=jnp.float32),
    key=jnp.array([0, 0], dtype=jnp.uint32),
)
next_state, _ = env.step(state, 3)
print(next_state.action_mask)
```
This code outputs `[True, True, False, False]`, but it should output `[False, False, False, False]`. In a related error, a random cell will be added even if the previous action failed.
```
env = Game2048()
state = State(
    board=jnp.array(
        [[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]], dtype=jnp.int32
    ),
    step_count=jnp.array(0, dtype=jnp.int32),
    action_mask=jnp.array([False, True, True, True], dtype=jnp.bool_),
    score=jnp.array(0.0, dtype=jnp.float32),
    key=jnp.array([0, 0], dtype=jnp.uint32),
)
next_state, _ = env.step(state, 0)
print(next_state.board)
```
This code outputs `[[1, 1, 1, 1], [0, 0, 1, 0], [0, 0, 0, 0], [0, 0, 0, 0]]`, but it should output `[[1, 1, 1, 1], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]`.

### What Jumanji version are you using?

jumanji 0.2.2

### Which accelerator(s) are you using?

CPU/GPU

### Additional System Info

3.11

### (Optional) Suggestion

The action_mask and done variables should be calculated after adding the random cell and the condition to add the random cell should be changed to `state.action_mask[action]`.

**Comment by clement-bonnet** [comment]
Hi @aar65537, thank you for raising this issue!
@OmaymaMahjoub, would you be able to take a look?
